core/src/main/java/org/bitcoinj/core/Wallet.java

```diff
@@ -1691,24 +1692,46 @@ private boolean checkForDoubleSpendAgainstPending(Transaction tx, boolean takeAc
             outpoints.add(input.getOutpoint());
         }
         // Now for each pending transaction, see if it shares any outpoints with this tx.
-        LinkedList<Transaction> doubleSpentTxns = Lists.newLinkedList();
-        for (Transaction p : pending.values()) {
+        Set<Transaction> doubleSpendTxns = Sets.newHashSet();
+        for (Transaction p : candidates.values()) {
             for (TransactionInput input : p.getInputs()) {
                 // This relies on the fact that TransactionOutPoint equality is defined at the protocol not object
                 // level - outpoints from two different inputs that point to the same output compare the same.
                 TransactionOutPoint outpoint = input.getOutpoint();
                 if (outpoints.contains(outpoint)) {
-                    // It does, it's a double spend against the pending pool, which makes it relevant.
-                    if (!doubleSpentTxns.isEmpty() && doubleSpentTxns.getLast() == p) continue;
-                    doubleSpentTxns.add(p);
+                    // It does, it's a double spend against the candidates, which makes it relevant.
+                    doubleSpendTxns.add(p);
                 }
             }
         }
-        if (takeAction && !doubleSpentTxns.isEmpty()) {
-            killTx(tx, doubleSpentTxns);
+        return doubleSpendTxns;
+    }
+
+    /**
+     * Adds to txSet all the txns in txPool spending outputs of txns in txSet, 
+     * and all txns spending the outputs of those txns, recursively.
+     * @param txSet The set of txns whose dependencies should be added. This method updates txSet.
+     * @param txPool The set of txns to evaluate adding to txSet
+     */
+    void addTransactionsDependingOn(Set<Transaction> txSet, Set<Transaction> txPool) {
+        Deque<Transaction> txDeque = new LinkedList<Transaction>();
+        txDeque.addAll(txSet);
+        while(!txDeque.isEmpty()) {
+            Transaction tx = txDeque.poll();
+            for (Transaction anotherTx : txPool) {
+                if (anotherTx.equals(tx)) continue;
+                for (TransactionInput input : anotherTx.getInputs()) {
+                    if (input.getOutpoint().getHash().equals(tx.getHash())) {
+                        if (!txDeque.contains(anotherTx)) {
```

Comment:
contains() in a LinkedList is quite expensive. Perhaps we want a LinkedHashMap here so we get fast random access? Or maybe it doesn't matter much.